File: Neuroflow/NeuralNetworks/RTLR.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Neuroflow.NeuralNetworks
{
    /// <summary>
    /// See:
    /// http://www.willamette.edu/~gorr/classes/cs449/rtrl.html
    /// </summary>
    internal sealed class RTLR : DisposableObject
    {
        internal RTLR(MultilayerPerceptron mlp)
        {
            Debug.Assert(mlp != null);

            this.mlp = mlp;
            this.netValueDerivates = mlp.AsMarshaled(mlp.NetValueDerivates.Values.ToArray());
            this.inputLayerInfos = mlp.AsMarshaled(
                (from lidx in Enumerable.Range(1, mlp.Layers.Count - 1)
                 let layer = mlp.Layers[lidx].Layer
                 select (from inputLayer in layer.GetInputLayers()
                         let iidx = mlp.GetLayerIndex(inputLayer)
                         select new RTLRLayerInfo
                         {
                             Index = iidx - 1,
                             Size = inputLayer.Size,
                             Weights = mlp.Weights[Tuple.Create(iidx, lidx)],
                             IsElementOfU = inputLayer != mlp.Layers[0].Layer
                         }).ToArray()).ToArray());

            CreatePValues(mlp);
        }

        MultilayerPerceptron mlp;

        IDeviceArrayPool pValuesPool;

        List<Marshaled<IDeviceArray[]>[][]> pWeightValues = new List<Marshaled<IDeviceArray[]>[][]>();

        List<Action<IDeviceArray, IDeviceArray>> codes = new List<Action<IDeviceArray, IDeviceArray>>();

        Marshaled<RTLRLayerInfo[][]> inputLayerInfos;

        Marshaled<IDeviceArray[]> netValueDerivates;

        private void CreatePValues(MultilayerPerceptron mlp)
        {
            pValuesPool = mlp.Adapter.DeviceArrayManagement.CreatePool();

            int uLayersCount = mlp.Layers.Count - 1;
            for (int lidx = 1; lidx < mlp.Layers.Count; lidx++)
            {
                var layer = mlp.Layers[lidx];
                var biases = mlp.Biases[lidx];

                var pWeightValues = new LinkedList<Marshaled<IDeviceArray[]>[]>();

                // Biases:
                var pWeightValuesOfInput = new Marshaled<IDeviceArray[]>[biases.Size];

                for (int biasIndex = 0; biasIndex < biases.Size; biasIndex++)
                {
                    pWeightValuesOfInput[biasIndex] = mlp.AsMarshaled(new IDeviceArray[uLayersCount]);

                    for (int lidx2 = 0; lidx2 < uLayersCount; lidx2++)
                    {
                        pWeightValuesOfInput[biasIndex].Instance()[lidx2] = pValuesPool.CreateArray(GetULayerSize(lidx2));
                    }
                }

                pWeightValues.AddLast(pWeightValuesOfInput);

                // Weighted conns:
                foreach (var inputConnectedLayer in layer.Layer.GetInputLayers())
                {
                    int inputIndex = mlp.GetLayerIndex(inputConnectedLayer);
                    var key = Tuple.Create(inputIndex, lidx);
                    var weigths = mlp.Weights[key];

                    pWeightValuesOfInput = new Marshaled<IDeviceArray[]>[weigths.Size];

                    for (int weightIndex = 0; weightIndex < weigths.Size; weightIndex++)
                    {
                        pWeightValuesOfInput[weightIndex] = mlp.AsMarshaled(new IDeviceArray[uLayersCount]);
                        for (int lidx2 = 0; lidx2 < uLayersCount; lidx2++)
                        {
                            pWeightValuesOfInput[weightIndex].Instance()[lidx2] = pValuesPool.CreateArray(GetULayerSize(lidx2));
                        }
                    }

                    pWeightValues.AddLast(pWeightValuesOfInput);
                }

                this.pWeightValues.Add(pWeightValues.ToArray());
            }
        }

        internal void ComputeGradients(IDeviceArray desiredOutputs)
        {
            int computationIndex = 0;
            var outputs = mlp.GetNetValues(mlp.Layers.Count - 1);

            for (int lidx = 1; lidx < mlp.Layers.Count; lidx++)
            {
                var layer = mlp.Layers[lidx];
                int layerOutputBufferSize = layer.Layer.Size;
                int iLayerIndex = lidx - 1;

                var layerPWeightValues = GetPWeights(iLayerIndex);

                for (int iValueIndex = 0; iValueIndex < layerOutputBufferSize; iValueIndex++)
                {
                    // i: iLayerIndex, iValueIndex

                    for (int jLayerIndex = 0; jLayerIndex < layerPWeightValues.Length; jLayerIndex++)
                    {
                        var layerRelatedPWeightBuffs = layerPWeightValues[jLayerIndex];

                        if (jLayerIndex == 0)
                        {
                            // Biased connections

                            var valueRelatedPBuffs = layerRelatedPWeightBuffs[iValueIndex];

                            // i: iLayerIndex, iValueIndex
                            // j: jLayerIndex, jValueIndex

                            bool isFirst = lidx == 1 && iValueIndex == 0;

                            ComputeGradients(computationIndex++, valueRelatedPBuffs, iLayerIndex, iValueIndex, jLayerIndex, -1, iValueIndex, outputs, desiredOutputs, isFirst ? SequenceMarker.Begin : SequenceMarker.Inner);
                        }
                        else
                        {
                            // Weighted connections
                            int inputLayerSize = layerRelatedPWeightBuffs.Length / layerOutputBufferSize;

                            Debug.Assert(layerOutputBufferSize * inputLayerSize == layerRelatedPWeightBuffs.Length);

                            for (int jValueIndex = 0; jValueIndex < inputLayerSize; jValueIndex++)
                            {
                                int ijValueIndex = GetWeightValueIndex(jValueIndex, iValueIndex, inputLayerSize);

                                var valueRelatedPBuffs = layerRelatedPWeightBuffs[ijValueIndex];

                                // i: iLayerIndex, iValueIndex
                                // j: jLayerIndex, jValueIndex

                                bool isLast =
                                    lidx == mlp.Layers.Count - 1 &&
                                    iValueIndex == layerOutputBufferSize - 1 &&
                                    jLayerIndex == layerPWeightValues.Length - 1 &&
                                    jValueIndex == inputLayerSize - 1;


                                ComputeGradients(computationIndex++, valueRelatedPBuffs, iLayerIndex, iValueIndex, jLayerIndex, jValueIndex, ijValueIndex, outputs, desiredOutputs, isLast ? SequenceMarker.End : SequenceMarker.Inner);
                            }
                        }
                    }
                }
            }
        }

        private void ComputeGradients(int computationIndex, Marshaled<IDeviceArray[]> valueRelatedPBuffs, int iLayerIndex, int iValueIndex, int jLayerIndex, int jValueIndex, int ijValueIndex, IDeviceArray outputs, IDeviceArray desiredOutputs, SequenceMarker seqMark)
        {
#if DEBUG
            int outputLayerIndex = valueRelatedPBuffs.Instance().Length - 1;
            Debug.Assert(outputLayerIndex == mlp.Layers.Count - 2);
            Debug.Assert(outputLayerIndex == mlp.Layers[mlp.Layers.Count - 1].Index - 1);
#endif

            if (codes.Count > computationIndex)
            {
                var code = codes[computationIndex];
                if (code != null) code(outputs, desiredOutputs);
            }
            else
            {
                codes.EnsureSize(computationIndex + 1);
                Action<IDeviceArray, IDeviceArray> code = null;

                bool forBias = jValueIndex == -1;
                var dataM = mlp.AsMarshaled(new RTLRComputationData());
                var data = dataM.ManagedObject;

                int iLayerIndexN = iLayerIndex + 1;
                var iLayer = mlp.Layers[iLayerIndexN];
                data.ILayerIndex = iLayerIndex;
                data.IValueIndex = iValueIndex;
                data.JLayerIndex = jLayerIndex;
                data.JValueIndex = jValueIndex;
                data.IJValueIndex = ijValueIndex;
                if (forBias)
                {
                    Debug.Assert(jLayerIndex == 0);
                    data.BiasGradients = mlp.GetBiasGradients(iLayerIndexN);
                    data.BiasGradientSums = mlp.GetBiasGradientSums(iLayerIndexN);
                }
                else
                {
                    Debug.Assert(jLayerIndex > 0);
                    var inputLayerOfILayer = iLayer.Layer.GetInputLayer(jLayerIndex - 1);
                    var inputLayerOfILayerIndex = mlp.GetLayerIndex(inputLayerOfILayer);
                    var weightKey = Tuple.Create(inputLayerOfILayerIndex, iLayerIndexN);
                    data.Inputs = () => mlp.GetNetValues(inputLayerOfILayerIndex);
                    data.Gradients = mlp.GetGradients(weightKey);
                    data.GradientSums = mlp.GetGradientSums(weightKey);
                }

                Debug.Assert(!(data.BiasGradients == null && data.BiasGradientSums == null && data.Gradients == null && data.GradientSums == null));

                var state = mlp.CreateComputationState();
                code = (os, dos) => mlp.Adapter.ComputeActivation.ComputeGradientsRTLR(state, inputLayerInfos, netValueDerivates, dataM, valueRelatedPBuffs, os, dos, seqMark);

                codes[computationIndex] = code;
                code(outputs, desiredOutputs);
            }
        }

        private Marshaled<IDeviceArray[]>[][] GetPWeights(int iLayerIndex)
        {
            return pWeightValues[iLayerIndex];
        }

        static int GetWeightValueIndex(int upperValueIndex, int lowerValueIndex, int lowerLayerSize)
        {
            return lowerValueIndex * lowerLayerSize + upperValueIndex;
        }       

        private int GetULayerSize(int lidx)
        {
            return mlp.Layers[lidx + 1].Layer.Size;
        }

        #region Zero

        internal void Zero()
        {
            pValuesPool.Zero();
        }

        #endregion

        #region Free

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);

            Free();
        }

        private void Free()
        {
            if (pValuesPool != null)
            {
                ResourceManager.Free(pValuesPool);
                foreach (var p in pWeightValues) Free(p);
                pValuesPool = null;
            }
        }

        private void Free(Marshaled<IDeviceArray[]>[][] p)
        {
            foreach (var ip in p) Free(ip);
        }

        private void Free(Marshaled<IDeviceArray[]>[] p)
        {
            foreach (var ip in p) Free(ip);
        }

        private void Free(Marshaled<IDeviceArray[]> p)
        {
            ResourceManager.Free(p.Instance());
        }

        #endregion
    }
}

File: NeuroflowLib/rtlr.cpp
#include "stdafx.h"
#include "rtlr.h"
#include "multilayer_perceptron.h"
#include "device_array.h"
#include "device_array2.h"
#include "layer.h"
#include "device_array2_group.h"
#include "device_array_group.h"
#include "computation_context.h"
#include "device_array_management.h"
#include "device_array_pool.h"
#include "rtlr_computation_data.h"
#include "compute_activation.h"

USING
namespace ph = std::placeholders;

void rtlr::initialize(multilayer_perceptron* mlp)
{
    assert(mlp);

    _mlp = mlp;
    _netValueDerivates = _mlp->_netValueDerivates.get_arrays() | to_vector();
    _pValuesPool = _mlp->context()->device_array_management()->create_pool(false);
    _uLayersCount = _mlp->_layers.size() - 1;
    _maxULayerSize = _mlp->_layers |
    where([](row_numbered<layer_ptr>& l) { return l.row_num() != 0; }) |
    max_value([](row_numbered<layer_ptr>& l) { return l.row_num(); });
    
    for (idx_t lidx = 1; lidx < _mlp->_layers.size(); lidx++)
    {
        _inputLayerInfos.emplace_back();
        _pValues.emplace_back();
        auto& layer = _mlp->_layers[lidx].value();
        _pValues.back().emplace_back(create_p_values_for_weights(_mlp->_biases.get(lidx)));
        for (auto& inputLayer : layer->input_layers())
        {
            idx_t iidx = _mlp->get_layer_index(inputLayer);
            _inputLayerInfos.back().emplace_back();
            auto& info = _inputLayerInfos.back().back();
            info.index = iidx - 1;
            info.size = inputLayer->size();
            info.weights = _mlp->_weights.get(make_pair(iidx, lidx));
            info.is_element_of_u = inputLayer != _mlp->_layers[0].value();
            _pValues.back().emplace_back(info.weights);
        }
    }
}

device_array2_ptr rtlr::create_p_values_for_weights(const device_array_ptr& weights)
{
    idx_t xSize = _uLayersCount * _maxULayerSize;
    idx_t ySize = weights->size();
    return _pValuesPool->create_array2(ySize, xSize);
}

void rtlr::compute_gradients(device_array* desiredOutputs)
{
    auto& outputs = _mlp->_netOutputs;
    idx_t computationIndex = 0;
    idx_t layersCount = _mlp->_layers.size();
    for (idx_t lidx = 1; lidx < layersCount; lidx++)
    {
        idx_t iLayerIndex = lidx - 1;
        auto& pValuesOfLayer = _pValues[iLayerIndex];
        idx_t pValuesOfLayersCount = pValuesOfLayer.size();
        for (idx_t jLayerIndex = 0; jLayerIndex < pValuesOfLayersCount; jLayerIndex++)
        {
            // 0: Bias
            // 1..: Weights
            auto& pValuesOfWeights = pValuesOfLayer[jLayerIndex];
            sequence_marker seqMark = sequence_marker::inner;
            if (lidx == 1 && jLayerIndex == 0) seqMark = sequence_marker::begin;
            else if (lidx == layersCount - 1 && jLayerIndex == pValuesOfLayersCount - 1) seqMark = sequence_marker::end;
            compute_gradients(iLayerIndex, jLayerIndex, pValuesOfWeights, outputs, desiredOutputs, computationIndex++, seqMark);
        }
    }
}

void rtlr::compute_gradients(
    idx_t iLayerIndex,
    idx_t jLayerIndex,
    const device_array2_ptr& pValuesOfWeights,
    device_array* outputs,
    device_array* desiredOutputs,
    idx_t computationIndex,
    sequence_marker seqMark)
{
    // jLayerIndex: 0: Bias, 1..: Weights

    if (_steps.size() <= computationIndex)
    {
        _steps.resize(computationIndex + 1);
        bool forBias = jLayerIndex == 0;
        rtlr_computation_data data;
        data.u_layers_count = _uLayersCount;
        data.max_u_layer_size = _maxULayerSize;
        idx_t iLayerIndexN = iLayerIndex + 1;
        auto& iLayer = _mlp->_layers[iLayerIndexN].value();
        data.i_layer_index = iLayerIndex;
        data.j_layer_index = jLayerIndex;
        if (forBias)
        {
            assert(jLayerIndex == 0);
            _mlp->_biasGradients.try_get(iLayerIndexN, data.bias_gradients);
            _mlp->_biasGradientSums.try_get(iLayerIndexN, data.bias_gradient_sums);
        }
        else
        {
            assert(jLayerIndex > 0);
            idx_t jLayerIndexN = jLayerIndex - 1;
            auto& inputLayerOfILayer = iLayer->get_input_layer(jLayerIndexN);
            idx_t inputLayerOfILayerIndex = _mlp->get_layer_index(inputLayerOfILayer);
            auto weightKey = make_pair(inputLayerOfILayerIndex, iLayerIndexN);
            data.inputs = [=]() { return _mlp->get_net_values(inputLayerOfILayerIndex); };
            _mlp->_gradients.try_get(weightKey, data.gradients);
            _mlp->_gradientSums.try_get(weightKey, data.gradient_sums);
        }

        assert(!(data.bias_gradients == null && data.bias_gradient_sums == null && data.gradients == null && data.gradient_sums == null));

        auto comp = _mlp->context()->compute_activation();
        _steps[computationIndex] =
        std::bind(
        [=](const nf_object_ptr& context, const rtlr_computation_data& data, device_array* outputs, device_array* desiredOutputs)
        {
            comp->compute_gradients_rtlr(
                context, 
                this->_inputLayerInfos, 
                this->_netValueDerivates, 
                data, 
                pValuesOfWeights, 
                outputs, 
                desiredOutputs, 
                seqMark);
        },
        move(comp->create_operation_context()),
        move(data),
        ph::_1,
        ph::_2);
    }

    _steps[computationIndex](outputs, desiredOutputs);
}